odl/solvers/functional/derivatives.py

```diff
@@ -0,0 +1,323 @@
+    """
+
+    def __init__(self, operator, point, method='forward', step=None):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        functional : `Operator`
+            The operator whose derivative should be computed. Its domain and
+            range must be `FnBase` spaces.
+        point : ``functional.domain`` `element-like`
+            The point to compute the derivative in.
+        method : {'backward', 'forward', 'central'}
+            The method to use to compute the derivative.
+        step : `None` or float
+            The step length used in the derivative computation.
+            `None` is interpreted as selecting the step according to the dtype
+            of the space. ``step = 1e-8`` for ``float64`` spaces and
+            ``step = 1e-4`` for ``float32`` spaces.
+
+        Examples
+        --------
+        Compute a numerical estimate of the derivative of the L2^2 norm:
+
+        >>> import odl
+        >>> space = odl.rn(3)
+        >>> func = odl.solvers.L2NormSquared(space)
+        >>> hess = NumericalDerivative(func.gradient, [1, 1, 1])
+        >>> hess([0, 0, 1])
+        rn(3).element([0.0, 0.0, 2.0])
+
+        Find the hessian matrix:
+
+        >>> odl.matrix_representation(hess)
+        array([[ 2.,  0.,  0.],
+               [ 0.,  2.,  0.],
+               [ 0.,  0.,  2.]])
+
+        Notes
+        -----
+        If the operator is :math:`A(x)` and stepsize :math:`h` is used, the
+        gradient is computed as follows.
+
+        If ``method='backward'`` is used:
+
+        .. math::
+            \\frac{A(x) - A(x - h)}{h}
+
+        if ``method='forward'`` is used:
+
+        .. math::
+            \\frac{A(x + h) - A(x)}{h}
+
+        if ``method='central'`` is used:
+
+        .. math::
+            \\frac{A(x + h/2) - A(x - h/2)}{h}
+
+        The number of function evaluations is ``functional.domain.size + 1`` if
+        ``'backward'`` or ``'forward'`` is used and
+        ``2 * functional.domain.size`` if ``'central'`` is used.
+        """
+        if not isinstance(operator, Operator):
+            raise TypeError('`operator` has to be a `Operator` instance')
+
+        if not isinstance(operator.domain, FnBase):
+            raise TypeError('`operator.domain` has to be a `FnBase` '
+                            'instance')
+        if not isinstance(operator.range, FnBase):
+            raise TypeError('`operator.range` has to be a `FnBase` '
+                            'instance')
+
+        self.operator = operator
+        self.point = operator.domain.element(point)
+
+        if step is None:
+            self.step = np.sqrt(np.finfo(operator.domain.dtype).eps)
+        else:
+            self.step = float(step)
+
+        self.method = str(method).lower()
+        if method not in ('backward', 'forward', 'central'):
+            raise ValueError('`method` not understood')
+
+        # TODO: should we really set linear to True here?
```

Comment:
> Write something about this choice in the class documentation, at the very top since it's important.

How about that? A single sentence would be enough.